Add keyflux plugin

salmanyam opened the pull request:
This plugin generates credentials (keys and certificates) for both the API proxy server (required for
kata-containers/kata-containers#9159 and
kata-containers/kata-containers#9752) and the workload owner. This plugin also delivers the credentials to a sandbox (i.e., confidential PODs or VMs), specifically to the kata agent to initiate the SplitAPI proxy server so that a workload owner can communicate with the proxy server using a secure tunnel.

The IPv4 address, name, and the ID of the sandbox (i.e., pod) must be provided in the query string to obtain the credentials from the KBS.

After receiving the credential request, the splitapi plugin will create a key pair for the server and client and sign them using the self-signed CA. The generated credentials are stored in a hashmap with a unique key for each sandbox based on its name, ID, and IP address. SplitAPI plugin responds to a request from the sandbox by sending the server specific credentials (key, cert) along with the CA certificate. A request from workload owner gets the client specific credentials (key, cert) and CA's certificate.

The splitapi plugin itself is not built or initialized by default. To build or initialize it, following steps need to be followed.

# How to build
1. Enable `splitapi-plugin` feature while building `kbs`.
2. Update the `kbs/config/kbs-config.toml` to enable the `splitapi` plugin.


```
$ make background-check-kbs POLICY_ENGINE=opa SPLITAPI_PLUGIN=true
$ cat >> kbs/config/kbs-config.toml << EOF
[[plugins]]
name = "pki_vault"
#plugin_dir = "/opt/confidential-containers/kbs/plugin/splitapi"
#cred_filename = "certificates.json"
[plugins.pkivault_cert_details]
country = "AA"
#state = "Default State"
#locality = "Default City"
organization = "Default Organization"
org_unit = "Default Unit"

[plugins.pkivault_cert_details.ca]
#common_name = "grpc-tls CA"
validity_days = 3650

[plugins.pkivault_cert_details.server]
#common_name = "server"
#validity_days = 180

[plugins.pkivault_cert_details.client]
common_name = "client"
validity_days = 180
```
 > **Note:** Only `name` is mandatory for the plugin configuration. The rest fields will get default values in case those are not configured in the `kbs-config.toml`.

# How to test
This plugin can be tested similarly to the `nebula-ca` plugin (#539).  We need the kbs-client patch available in the branch https://github.com/cclaudio/trustee/tree/nebula-ca-plugin-test to build the kbs-client and test the plugin.

Once the `kbs-client` is built, use the `kbs-client` to make a request to KBS for generating and providing the server specific credentials.

```
$ cd kbs && make cli ATTESTER=snp-attester
$ sudo make install-cli
$ kbs-client --url http://127.0.0.1:8080 get-resource --plugin-name "pki_vault" --resource-path "credential?id=3367353&ip=60.11.12.48&name=pod33" | base64 -d
```
KBS should return the server specific credentials such as server key, server certificates, and the certificate of the CA.

salmanyam committed:
Add keyflux plugin for dynamic secret generation

The keyflux plugin includes the following features:
- generation of tls, symmetric, ed25519, and rsa secrets
- apis for server and client to request the secrets
- apis for updating server's or client's tls cert details
- configuration for configuring CA and secrets

Signed-off-by: Salman Ahmed <sahmed@ibm.com>

bpradipt commented:
> This plugin generates credentials (keys and certificates) for both the API proxy server (required for [kata-containers/kata-containers#9159](https://github.com/kata-containers/kata-containers/pull/9159) and [kata-containers/kata-containers#9752](https://github.com/kata-containers/kata-containers/pull/9752)) and the workload owner. This plugin also delivers the credentials to a sandbox (i.e., confidential PODs or VMs), specifically to the kata agent to initiate the SplitAPI proxy server so that a workload owner can communicate with the proxy server using a secure tunnel.
> 
> The IPv4 address, name, and the ID of the sandbox must be provided in the query string to obtain the credential resources from the kbs.
> 
> After receiving the credential request, the splitapi plugin will create a key pair for the server and client and sign them using the self-signed CA. The generated ca.crt, server.crt, and server.key are stored in a directory specific to the sandbox (the caller) and returned to the caller. In addition, ca.key, client.key, and client.crt are also generated and stored to that particular directory specific to the sandbox (i.e., caller).
> 
> During the credential generation, a sandbox directory mapper creates a unique directory specific to the sandbox (i.e., the caller). The mapper creates the unique directory using the sandbox parameters passed in the query string. A mapping file is also maintained to store the mapping between the sandbox name and the unique directory created for the sandbox.
> 
> The splitapi plugin itself is not initialized by default. To initialize it, you need to add 'splitapi' in the kbs-config.toml.

A generic question. Would it make sense to have a generic plugin to create key-pair for a sandbox and not club it with splitapi?
There are other use cases for per sandbox (pod or VM) keys especially in the peer-pods case.
cc @yoheiueda @davidhadas

fitzthum commented:
> A generic question. Would it make sense to have a generic plugin to create key-pair for a sandbox and not club it with splitapi?

Some kind of PKI keypair would be very useful.

salmanyam commented:
> A generic question. Would it make sense to have a generic plugin to create key-pair for a sandbox and not club it with splitapi?
> There are other use cases for per sandbox (pod or VM) keys especially in the peer-pods case.

We are good. I think it would be a good idea to have the key generation as a basic service needed by other plugins. We can definitely consider refactoring and making it as a basic service. In that case, the splitapi plugin will utilize this basic service.

yoheiueda commented:
A generic PKI approach will be very helpful for our peer pod use cases.

I I understand correctly, the current proposal is to run a key generation service in a trustee, but a PKI approach sounds to me  a signing service in a trustee.  

Trustee has a private key for CA, and accepts key signing requests from clients. A client in a sandbox generates a key pair, and send a key signing request with the generated public key. The CA in Trustee issues a certificate by singing the received public key using the CA private key, and distribute the issued certificate. I think this is how the current PKI for TLS certificates works, and the sandbox does not need to reveal its private key to Trustee.

I think this will work If the purpose of the certificate generation is TLS secure tunneling.

This is just out of my curiosity, and  a key generation service will still very helpful for our peer pod use case.

Another point is that the current proposal generates RSA keys. To support other encryption mechanisms such as WireGuard and SSH, it is helpful to support different types of keys such as Ed25529. WireGuard uses Curve25529 keys, and Ed25529 is also prefabbed for SSH these days.

fitzthum commented on kbs/src/plugins/implementations/splitapi/manager.rs:
```diff
@@ -0,0 +1,163 @@
+                sandbox_dir_info = existing_dir.clone();
+
+                //TODO: check if the credentails are already in there
+                // send the existing credentials if they are not expired
```
When you create the certs, just store the expiration time alongside them so you can easily see if they are expired.

fitzthum commented on kbs/src/plugins/implementations/splitapi/mapper.rs:
I think it's a mistake to store all this stuff on the filesystem.

Here's what I would do. Create a struct that contains all the state that represents a keypair. OpenSSL has structs for all the things you need. Store this struct in a dictionary keyed by the connection id. Wrap the struct in a RwLock. If you need some kind of persistence you can serialize/desrialize the dictionary as needed.

I don't see any reason to be writing things to the filesystem and I think this entire mapper concept can be replaced by something from std, which decreases your complexity significantly.

fitzthum commented on kbs/src/plugins/implementations/splitapi/manager.rs:
```diff
@@ -0,0 +1,163 @@
+        if let Some(mapper) = mapper_guard.as_mut() {
+            let sandbox_dir_info: SandboxDirectoryInfo;
+
+            if let Some(existing_dir) = mapper.get_directory(&params.name) {
```
These params are sent in other the network in plaintext. Would it be a problem if someone tampered with the name value in flight?

More fundamentally, why does this need to be stateful at all?

fitzthum reviewed (commented):
Made a few comments. Please run `cargo fmt` and `cargo clippy` to fix the CI.

Xynnn007 commented on kbs/src/plugins/implementations/splitapi/backend.rs:
```diff
@@ -0,0 +1,77 @@
+
+#[derive(Clone, Debug, Deserialize, PartialEq)]
+#[serde(tag = "type")]
+pub enum SplitAPIConfig {
+    CertManager(manager::SplitAPIRepoDesc),
+}
```
If we use `enum` here, we are hinting that potentially we could have other types of `SplitAPIConfig` besides `CertManager`. Do we? If not, I suggest to change `SplitAPIConfig` into `struct` and move members of `SplitAPIRepoDesc` directly into `SplitAPIConfig`

Xynnn007 commented on kbs/src/plugins/implementations/splitapi/manager.rs:
```diff
@@ -0,0 +1,163 @@
+
+
+// Use lazy_static to initialize the SANDBOX_DIRECTORY_MANAGER only once
+lazy_static! {
+    static ref SANDBOX_DIRECTORY_MAPPER: Arc<Mutex<Option<SandboxDirectoryMapper>>> = Arc::new(Mutex::new(None));
+}
```
Could we move this static variable directly into `CertManager` as a member? s.t.
```rust
pub struct CertManager {
    pub plugin_dir: String,
    pub mapping_filename: String,
    mapper: Arc<Mutex<SandboxDirectoryMapper>>,
}
```

Also, could we use [tokio::sync::Mutex](https://docs.rs/tokio/1.41.1/tokio/sync/struct.Mutex.html#examples) rather than `std::sync::Mutex`? The async version would promote CPU efficiency.

Xynnn007 commented on kbs/src/plugins/implementations/splitapi/manager.rs:
```diff
@@ -0,0 +1,163 @@
+        // Try locking the sandbox directory mapper
+        let mut mapper_guard = self.mapper.lock().map_err(|e| {
+            anyhow!("Failed to lock sandbox directory mapper: {}", e)
+        })?;
```
`anyhow` provides an easier way for `map_err(|e| anyhow!(..., e))` like things
```rust
xxx.context("Failed to lock sandbox directory mapper")?;
```

salmanyam commented:
Hi @fitzthum and @Xynnn007, I have addressed your comments in the latest commit. Could you please take a look? Thank you!

fitzthum commented on kbs/src/plugins/implementations/splitapi/backend.rs:
```diff
@@ -0,0 +1,83 @@
+use super::generator::CertificateDetails;
+use super::manager;
+
+pub const CREDENTIALS_BLOB_FILE: &str = "certificates.json";
```
You're using the term `credential` throughout the PR, but also the term certificates. It might be clearer to just use one especially on lines like this one.

fitzthum commented on kbs/src/plugins/implementations/splitapi/manager.rs:
```diff
@@ -0,0 +1,148 @@
+
+        log::info!("Credentials are generated!");
+
+        // Aquire the write lock and write the credential into the hashmap
```
It's a little odd to be writing state in a method called `get...`

fitzthum commented on kbs/src/plugins/implementations/splitapi/manager.rs:
```diff
@@ -0,0 +1,148 @@
+
+#[async_trait::async_trait]
+impl SplitAPIBackend for CertManager {
+    async fn get_server_credential(&self, params: &SandboxParams) -> Result<Vec<u8>> {
```
What is the `server` here? These credentials are for the client, right? Maybe you're referring to some other server, like the SplitAPI server in the guest, but this is pretty confusing.

fitzthum commented on kbs/src/plugins/implementations/splitapi/manager.rs:
```diff
@@ -0,0 +1,148 @@
+        })
+    }
+
+    async fn load_credentials(&self, key: &str) -> Option<Credentials> {
```
nit: you are alternating between `credentials` (plural) and `credential` singular in your method names. Probably best to pick one.

Also, this function should probably return `Result<Option<Credentials>>` given that you could have a failure to load the credentials (i.e. if the file doesn't exist or something) or you could legitimately return no credentials.

fitzthum commented on kbs/src/plugins/implementations/splitapi/manager.rs:
```diff
@@ -0,0 +1,148 @@
+
+    async fn load_credentials(&self, key: &str) -> Option<Credentials> {
+        // Check if the credential is not loaded. If not, load them
+        if !self.credential_loaded_from_file.load(Ordering::SeqCst) {
```
Can we just do this at initialization time and avoid keeping track of whether they've been loaded with every request?

salmanyam replied:
@fitzthum, Thank you very much for providing feedback.

The reason for doing the loading this way is because the plugin initialization code (in `plugin_manager.rs`), that initializes the plugins, is synchronous, but the `load_credentials` function is async. Since the `async` function cannot be used inside a `sync` function, I cannot directly call the `load_credentials` function when the `splitapi` plugin is initialized. There might be two ways to solve it:

1. changing the `load_credentials` function to be synchronous, or
2. making the plugin initialization code to be asynchronous

I am not sure if there's a better way to call an async function inside a sync function.

fitzthum replied:
Either one seems fine. Probably easier to make this sync. You're just reading from a file so that should be doable. I think you can make a new `tokio::sync::arc` without async.

fitzthum reviewed (commented):
Thanks for the updates. I think this is a big improvement. At some point you should squash your commits together. I made some comments.

bpradipt commented:
> > A generic question. Would it make sense to have a generic plugin to create key-pair for a sandbox and not club it with splitapi?
> 
> Some kind of PKI plugin would be very useful.

@salmanyam any plans to revisit this PR from the pov of creating a generic PKI plugin ?

salmanyam commented:
> > > A generic question. Would it make sense to have a generic plugin to create key-pair for a sandbox and not club it with splitapi?
> > 
> > 
> > Some kind of PKI plugin would be very useful.
> 
> @salmanyam any plans to revisit this PR from the pov of creating a generic PKI plugin ?

Hi @bpradipt, yes, I almost finished working on that aspect of the PR. The redesign of the PR will not be tied with splitapi. I will update the PR within a couple days.

portersrc commented:
Hey @salmanyam, cool work! One comment related to the `cred_store`:

Do you really need to maintain state with this plugin? You do it in two places. One is a hashmap called `cred_store`; and then you also persist that hashmap to file. The key to that hashmap is this IP-name-podID that you're using. Does a client ever need to request its credentials more than once? (Even if the pod dies and needs to get credentials again, won't it start up with a new podID, in which case it would need new keys anyway?)

salmanyam commented:
> Hey @salmanyam, cool work! One comment related to the `cred_store`:
> 
> Do you really need to maintain state with this plugin? You do it in two places. One is a hashmap called `cred_store`; and then you also persist that hashmap to file. The key to that hashmap is this IP-name-podID that you're using. Does a client ever need to request its credentials more than once? (Even if the pod dies and needs to get credentials again, won't it start up with a new podID, in which case it would need new keys anyway?)

Thanks, @portersrc! You brought a valid point. We also discussed this point in our internal meeting. Your are right. The client (or pod or sandbox) requests its credential only one time. But we need the persistence state for the following reason. That is if trustee restarts after the pod initialization, in that case, pod (client or sandbox) gets its credential, but the workload owner lose the CA that signs the pod's credential. As a result, the workload owner cannot perform mutual authentication between the pod and itself as we need the same CA for mutual authentication. Hope I was able to clarify.

portersrc commented:
I see. Is this PR as-is supporting this, though, or do you have more changes planned? I noticed when a new set of credentials have to be created, a new `ca_cert` is created. (That is, you don't sign each new `server_cert` with a common `ca_cert`; rather, you generate a new `ca_cert` for every new `server_cert`.)

Not sure if the following is related: The client credentials, though they get stored in a hashmap and persisted, aren't externally accessible. Maybe the client credentials need to have an API endpoint, as well, if you want any mutual authentication with the `ca_cert` that's common between the 1 server and 1 client?

salmanyam commented:
> I see. Is this PR as-is supporting this, though, or do you have more changes planned? I noticed when a new set of credentials have to be created, a new `ca_cert` is created. (That is, you don't sign each new `server_cert` with a common `ca_cert`; rather, you generate a new `ca_cert` for every new `server_cert`.)
> 
> Not sure if the following is related: The client credentials, though they get stored in a hashmap and persisted, aren't externally accessible. Maybe the client credentials need to have an API endpoint, as well, if you want any mutual authentication with the `ca_cert` that's common between the 1 server and 1 client?

The PR does not completely support whatever I said in the previous comment. I am working on the admin (workload owner) APIs. The admin APIs endpoints will allow workload owner to fetch the `client` credentials for the mutual authentication. Yes, to simplify the design, we decided to use a separate CA for each pod. The pod credentials can be obtained through the same API (`get_resource`).

bpradipt commented:
> @fitzthum, I have addressed your comments and refactored and cleaned the code. I have renamed the plugin name to `pki-vault`. I am also working to add the admin APIs which I will update soon. I would greatly appreciate your feedback.
> 
> Hi @bpradipt, The plugin name is now `pki-vault`. Currently, it generates all the keys [CA, server (sandabox), client (owner or admin)] and then generates the certificates for server and client signed by the CA certificate. I am working to support functionality where the sandbox or client (owner or admin) will generate their own key-pair and send the certificate signing request (CSR) to the plugin to sign it by the same CA. One key aspect of the current design is mutual authentication, i.e., both the sandbox and owner confirm each other. BTW, can you please tell me your use case? That will help me greatly for generalizing the design.

Thanks for working on this @salmanyam .. Following are my use cases:
1. Mutual authentication between the user (owner) and the CoCo pod
2. Mutual authentication between between two CoCo pods 
3. Mutual authentication between between a workload in the trusted env and CoCo pod (similar to 1)
4. A server in the trusted env (having the CA cert from Trustee) verifying the client certs sent from CoCo pods

Ability to send a CSR to the plugin from the pod as well as an API to retrieve the ca cert will be important imho. The features can be iteratively implemented.

Can you add a design doc or readme to the PR. It'll help to review.

fitzthum commented:
I agree with most of the things mentioned here. It seems like it would be simpler to not maintain any state. I was picturing there being one root cert, corresponding to Trustee itself. It seems like in the current implementation different pods would have different root certs, which is interesting.

Also, I'm not exactly sure about the data that is being provided from the client. The pod id and such. This seems a little arbitrary and it doesn't seem like this information actually makes it into the cert anyway. Do we want the cert to be endorsing something about the guest? One possible approach would be to allow clients to specify any values they want as part of a query string. This query string would pass through the policy engine, so Trustee admins could add their own logic about what is allowed. Then maybe these claims would be added to the cert as an extension or something.

Ultimately I'm not sure exactly what PKI for Trustee should look like, so if you have a vision here, feel free to go that direction. I think a design doc would help understand the sort of use cases that you are looking at.

salmanyam commented:
> > @fitzthum, I have addressed your comments and refactored and cleaned the code. I have renamed the plugin name to `pki-vault`. I am also working to add the admin APIs which I will update soon. I would greatly appreciate your feedback.
> > Hi @bpradipt, The plugin name is now `pki-vault`. Currently, it generates all the keys [CA, server (sandabox), client (owner or admin)] and then generates the certificates for server and client signed by the CA certificate. I am working to support functionality where the sandbox or client (owner or admin) will generate their own key-pair and send the certificate signing request (CSR) to the plugin to sign it by the same CA. One key aspect of the current design is mutual authentication, i.e., both the sandbox and owner confirm each other. BTW, can you please tell me your use case? That will help me greatly for generalizing the design.
> 
> Thanks for working on this @salmanyam .. Following are my use cases:
> 
> 1. Mutual authentication between the user (owner) and the CoCo pod
> 2. Mutual authentication between between two CoCo pods
> 3. Mutual authentication between between a workload in the trusted env and CoCo pod (similar to 1)
> 4. A server in the trusted env (having the CA cert from Trustee) verifying the client certs sent from CoCo pods
> 
> Ability to send a CSR to the plugin from the pod as well as an API to retrieve the ca cert will be important imho. The features can be iteratively implemented.
> 
> Can you add a design doc or readme to the PR. It'll help to review.

Hi @bpradipt, thank you for providing the use cases. It will help me greatly to design the admin (workload owner) APIs. I am working on the owner side APIs and design doc. I will update soon.

zvonkok commented:
/cc @zvonkok

bpradipt commented:
@salmanyam any updates on this ?

salmanyam commented:
> @salmanyam any updates on this ?

@bpradipt Yes, I have made progress (both in code and documentation). I will update what I have so far by today.

salmanyam commented:
Hi @bpradipt and @fitzthum, I have made progress to this work as it was sidelined for a few weeks due to other work. I have also started the documentation. Could you please take a look? Thank you!

fitzthum reviewed (commented):
Nice update. Mainly commented on the doc. Code looks tighter. An integration test would be nice. I think we are definitely moving in the right direction here.

bpradipt commented on kbs/docs/plugins/pki_vault.md:
````diff
@@ -0,0 +1,147 @@
+```toml
+[[plugins]]
+name = "pki_vault"
+plugin_dir = "/opt/confidential-containers/kbs/plugin/splitapi"
````
nit: the dir name splitapi doesn't align with the intended functionality here.

salmanyam committed:
Update doc of keyflux plugin

Signed-off-by: Salman Ahmed <sahmed@ibm.com>

salmanyam commented:
Hi @fitzthum and @Xynnn007 

I have significantly restructured the plugin according to your suggestions. I have also tested the plugin with the latest plugin-calling changes made by @Xynnn007. Could you please take a look at the latest changes? I have also rewritten the readme doc for the plugin. I have renamed the plugin name from `pki_vault` to `keyflux` to have the name aligned with what the plugin does.

## Key changes are the following.

- **Plugin name**: Changed from `pki_vault` to `keyflux`
- **Configuration structure**: Updated to use nested TOML sections
- **API paths**: Changed from `/kbs/v0/pki_vault/*` to `/kbs/v0/keyflux/*` (unauthenticated) and `/kbs/v0/keyflux/*` (authenticated)
- **Additional secret types**: Added support for Ed25519 and RSA keys beyond TLS and symmetric keys
- **Flexible query parameters**: Configurable required parameters instead of hardcoded `token`, `name`, `ip`

Please take a look the readme (kbs/docs/plugins/keyflux.md) for more details

fitzthum commented:
Thanks for the updates @salmanyam I'll review on monday

Xynnn007 commented on kbs/docs/plugins/keyflux.md:
```diff
@@ -0,0 +1,373 @@
+**Limits Configuration** (`plugins.keyflux.limits`):
+- `symmetric_key_size`: Size of symmetric keys in bytes (default: 32)
+- `rsa_bits`: RSA key size in bits (default: 2048)
+- `allow_types`: Allowed secret types (default: `["tls", "symmetric", "ed25519", "rsa"]`)
```
Do we have a list of `allow_types`?

Xynnn007 replied:
Oh I see the full list are the four. Let's explicitly show it

Xynnn007 commented on kbs/src/plugins/implementations/keyflux.rs:
```diff
@@ -0,0 +1,867 @@
+pub struct LimitsConfig {
+    pub symmetric_key_size: usize,
+    pub rsa_bits: usize,
+    pub allow_types: Vec<String>,
```
Looks like this field is never used?

Xynnn007 commented on kbs/src/plugins/implementations/keyflux.rs:
```diff
@@ -0,0 +1,867 @@
+            }
+        }
+    }
+}
```
This can be simplifiled by a `default` macro upon `KeyFluxPluginConfig`

Xynnn007 commented on kbs/src/plugins/implementations/keyflux.rs:
```diff
@@ -0,0 +1,867 @@
+
+            ed25519_store: Arc::new(RwLock::new(HashMap::new())),
+            rsa_store: Arc::new(RwLock::new(HashMap::new())),
+            symkey_store: Arc::new(RwLock::new(HashMap::new())),
```
We can use `tokio::sync::Rwlock` than `std`'s here. As the whole runtime is in the async

Xynnn007 commented on kbs/src/plugins/implementations/keyflux.rs:
```diff
@@ -0,0 +1,867 @@
+            let parts: Vec<&str> = item.split(':').collect();
+
+            if parts.len() != 2 {
+                return Err(anyhow!("Invalid spec format: {}", item));
```
```suggestion
                bail!("Invalid spec format: {}", item);
```

Xynnn007 commented on kbs/src/plugins/implementations/keyflux.rs:
```diff
@@ -0,0 +1,867 @@
+                "sym" | "symmetric" => SecretType::Symmetric,
+                "pkey" | "ed25519" => SecretType::Ed25519,
+                "rsa" => SecretType::Rsa,
+                other => return Err(anyhow!("Unknown secret type: {}", other)),
```
```suggestion
                other => bail!("Unknown secret type: {}", other),
```

Xynnn007 commented on kbs/src/plugins/implementations/keyflux.rs:
```diff
@@ -0,0 +1,867 @@
+                .map_err(|_| anyhow!("Invalid count for {}", parts[0]))?;
+
+            if count == 0 {
+                return Err(anyhow!("Count must be > 0 for {}", parts[0]));
```
```suggestion
                bail!("Count must be > 0 for {}", parts[0]);
```

Xynnn007 commented on kbs/src/plugins/implementations/keyflux.rs:
```diff
@@ -0,0 +1,867 @@
+    pub rsa_store: Arc<RwLock<HashMap<String, HashMap<String, Vec<u8>>>>>,
+
+    // secret stored for clients
+    pub symkey_store: Arc<RwLock<HashMap<String, HashMap<String, Vec<u8>>>>>,
```
Not the issue for this PR, but the KBS design. This HashMap will bring state to a single KBS instance and will be against https://github.com/confidential-containers/trustee/issues/1092

This is a design problem of KBS that a unified key value storage interface is not easily achieved. Let me mark this as a TODO for me.

fitzthum commented on kbs/docs/plugins/keyflux.md:
```diff
@@ -0,0 +1,373 @@
+**Endpoint**: `GET /kbs/v0/keyflux/credentials`
+
+**Query Parameters**:
+- `id` (required): Unique identifier for the pod
```
It might be good to note that this pod id is simply whatever the guest chooses. It isn't bound to the true pod id, or is it?

fitzthum commented on kbs/docs/plugins/keyflux.md:
````diff
@@ -0,0 +1,373 @@
+**Example Request**:
+
+```bash
+../target/release/kbs-client \
````
Slightly confusing example. You talk about this being used by pods, but you show the KBS client, which should not be used from within a pod. Maybe you want to show with the ASR rest API instead.

More generally, I'm not sure that this doc should be talking about pods. Trustee can be used outside of confidential containers. It seems like this plugin would also work with VMs. So maybe you want to get rid of the terms pod and sandbox entirely and give an example of getting credentials either via the rest API for coco or the kbspclient in a VM.

fitzthum commented on kbs/docs/plugins/keyflux.md:
````diff
@@ -0,0 +1,373 @@
+```bash
+../target/release/kbs-client \
+    --url http://localhost:8090 \
+    get-resource --path "credentials?id=pod123&spec=tls:1;sym:2;ed25519:1"
````
fwiw I don't really see the value in being able to request multiple keys at once. This makes things more complicated all around. If a user needs multiple keys they could just make multiple requests. I guess it is fine to keep this, tho.

fitzthum commented on kbs/docs/plugins/keyflux.md:
````diff
@@ -0,0 +1,373 @@
+```bash
+../target/release/kbs-client \
+    --url http://localhost:8090 \
+    config --auth-private-key config/private.key \
````
See recent changes to auth. We use a token here instead now.

fitzthum commented on kbs/docs/plugins/keyflux.md:
```diff
@@ -0,0 +1,373 @@
+
+6. **Establish mutual TLS connection** between pod VM (server) and workload owner (client) using the credentials
+
+## Security Considerations
```
this section is unnecessary imo

fitzthum commented on kbs/docs/plugins/keyflux.md:
```diff
@@ -0,0 +1,373 @@
+- **Authentication Required**: Client APIs require KBS private key authentication
+- **Automatic Encryption**: Server credentials are automatically encrypted with TEE public key
+
+## Troubleshooting
```
not sure this is needed either. The first issue is obvious. For the others, just add a bit of the trouble shooting info to the error message itself.

fitzthum commented on kbs/src/plugins/implementations/keyflux.rs:
```diff
@@ -0,0 +1,867 @@
+    fn default() -> Self {
+        Self {
+            keyflux: KeyFluxSection {
+                //secrets: vec![],
```
remove if not needed

fitzthum reviewed (commented):
I think this is an improvement. Made a few comments.

Btw, where does the term keyflux come from. I see there is a company and a rust crate already called keyflux, but they seem unrelated. If this is a new term you are coming up with, I might opt for something more descriptive.

salmanyam committed:
rename to secretprov and improve plugin design

- Rename plugin from keyflux to secretprov across all files (Cargo.toml,
  Makefile, source, docs); update all type names accordingly
- Stop storing private keys after TEE delivery: KeyFluxEntry now holds only
  public material (CA key retained for TLS client cert signing on demand)
- Remove one-set-only guard, has_entry/store_entry helpers, and refresh=true
  param; GET /credentials now regenerates unconditionally
- Switch std::sync::RwLock to tokio::sync::RwLock throughout
- Fix generate_ca_cert missing set_version(2) (X.509 v3 required by RFC 5280)
- Fix P256 branch to respect server_cert_config_store instead of hardcoding CN
- Remove unused constants DEFAULT_SERVER_VALIDITY_DAYS, DEFAULT_CLIENT_VALIDITY_DAYS,
  CA_DEFAULT_COMMON_NAME
- Update docs: remove pod-centric language, fix config field names
  (spec_required, supported_types), remove Security Considerations and
  Troubleshooting sections, replace kbs-client VM examples with REST API form

Signed-off-by: Salman Ahmed <sahmed@ibm.com>

salmanyam commented:
Hi @Xynnn007 and @fitzthum,

Sorry for the delay. I tried to address all comments. I also made some design and code quality improvements. Could you please take a look? Thank you!

@Xynnn007
- Replaced std::sync::RwLock with tokio::sync::RwLock throughout, converting all lock acquisitions to .read().await / .write().await. This avoids blocking the async runtime on lock contention.
- Removed the hand-written Default for KeyFluxPluginConfig
- Simplified the logic by removing the feature that generates multiple secrets at a time. Now, we generate one secret at a time which is much simpler
- Simplified update_cert_details — replaced the redundant and_modify(...).or_insert(...) pattern with a single insert call.

@fitzthum 
- Removed the unused constants DEFAULT_SERVER_VALIDITY_DAYS, DEFAULT_CLIENT_VALIDITY_DAYS, and CA_DEFAULT_COMMON_NAME.
- Removed the Security Considerations section from the documentation.
- Removed the Troubleshooting section from the documentation 
- Replaced pod/sandbox-centric language in the documentation with VM-centric language. The plugin is not specific to Kubernetes pods.
- Replaced the kbs-client examples for server-side credential requests with the KBS REST API form, since kbs-client should not be used from inside a confidential VM.
- Added a note that identity parameter values are supplied by the guest itself and are not verified against any external source — policy enforcement is delegated to the KBS resource policy.
- Updated config documentation to reflect the current field names (spec_required, supported_types).

Xynnn007 commented on kbs/config/kbs-config.toml:
```diff
@@ -54,3 +54,26 @@ storage_backend_type = "kvstorage"
 # path "secret/*" {
 #   capabilities = ["create", "read", "update", "delete", "list"]
 # }
+
+[[plugins]]
+name = "secretprov"
+
+[plugins.secretprov.ca]
+country = "AA"
+state = "Default State"
+locality = "Default City"
+organization = "Default Organization"
+org_unit = "Default Unit"
+common_name = "grpc-tls CA"
+validity_days = 3650
+
+[plugins.secretprov.query]
+# Identity params joined to build the store key, e.g. "pod-abc_default"
+required = ["name", "ns"]
+# Require secret_name + secret_type on every request
+spec_required = true
+
+[plugins.secretprov.limits]
+symmetric_key_size = 32
+rsa_bits = 2048
+supported_types = ["tls", "symmetric", "ed25519", "rsa", "p256"]
```
Let's not add the configs here as it's used for docker-compose deployments and will not take effect.

salmanyam replied:
Sounds good. Will remove this.

fitzthum commented on kbs/src/plugins/implementations/secretprov.rs:
```diff
@@ -0,0 +1,876 @@
+pub const DEFAULT_STATE: &str = "Default State";
+pub const DEFAULT_LOCALITY: &str = "Default City";
+pub const DEFAULT_ORGANIZATION: &str = "Default Organization";
+pub const DEFAULT_ORG_UNIT: &str = "Default Unit";
```
maybe these defaults should mention coco/trustee

fitzthum commented on kbs/src/plugins/implementations/secretprov.rs:
```diff
@@ -0,0 +1,876 @@
+pub const DEFAULT_LOCALITY: &str = "Default City";
+pub const DEFAULT_ORGANIZATION: &str = "Default Organization";
+pub const DEFAULT_ORG_UNIT: &str = "Default Unit";
+pub const DEFAULT_CA_VALIDITY_DAYS: u32 = 3650;
```
Seems a little high. These certs are one-per-guest, right?

fitzthum commented on kbs/src/plugins/implementations/secretprov.rs:
```diff
@@ -0,0 +1,876 @@
+
+/// Resource limits applied during secret generation.
+#[derive(Clone, Debug, Deserialize, PartialEq)]
+pub struct LimitsConfig {
```
These aren't really limits. They're just settings.

fitzthum commented on kbs/src/plugins/implementations/secretprov.rs:
```diff
@@ -0,0 +1,876 @@
+    /// Handles `GET /credentials`.
+    ///
+    /// Generates a fresh key pair on every call and overwrites any previously
+    /// stored public material for the same (identity-key, spec-sub-key) pair.
```
It's interesting that you're overwriting an existing key. Since the identity is derived from the query string, which is unmeasured, could a third party overwrite someone else's entry?

salmanyam replied:
yes, that's a possibility. Do you have any suggestions or pointers here?

fitzthum replied:
Well you could get these parameters from init-data rather than from the query string. This would be bound to the Tee instance.

If you don't overwrite the connections, you risk mixing public keys between guests.

You could potentially stick with the current approach, but you probably want to document the limitations. Speaking of which, this can't really support certs that expire, right?

salmanyam replied:
I like the init-data approach. I would like to give it a try. Regarding the certs expiration, yes, I just realized that. Let me think here and come up what is better.

fitzthum replied:
Check out https://github.com/confidential-containers/trustee/pull/1305 regarding init-data. That plugin is relatively similar to this one, but it expects an out-of-band registration operation and it's more tailored to LUKs keys.

fitzthum commented on kbs/src/plugins/implementations/secretprov.rs:
```diff
@@ -0,0 +1,876 @@
+            store: Arc::new(RwLock::new(HashMap::new())),
+
+            server_cert_config_store: Arc::new(RwLock::new(HashMap::new())),
+            client_cert_config_store: Arc::new(RwLock::new(HashMap::new())),
```
Well, at this point we should really be using the kvstorage backend for this.

fitzthum commented on kbs/src/plugins/implementations/secretprov.rs:
```diff
@@ -0,0 +1,876 @@
+/// `spec_required` gates whether `secret_name` and `secret_type` must appear
+/// on every request (default: true).
+#[derive(Clone, Debug, Deserialize, PartialEq)]
+pub struct QueryConfig {
```
Do you really need to have a configurable mapping for this? Seems like it would be simpler to just have one set of required args, at least in the first iteration.

fitzthum commented on kbs/src/plugins/implementations/secretprov.rs:
```diff
@@ -0,0 +1,876 @@
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Default)]
+pub struct SecretProvSection {
```
nit: `section` is kind of a vague term.

fitzthum reviewed (commented):
A couple comments, but this is looking pretty good.

tbh I'm not sure `SecretProv` is the best name either. All of the plugins provision secrets. This one is more like a keypair generator?

salmanyam commented:
All these comments make sense. I will address them asap.
